src/future/flatten_sink.rs

```diff
@@ -0,0 +1,91 @@
+use std::mem;
+
+use {Async, AsyncSink, StartSend, Poll, Future, Sink};
+
+
+enum State<F: Future>
+    where F::Item: Sink
+{
+    Future(F),
+    Sink(F::Item),
+    Error(<F::Item as Sink>::SinkError),
+    Void,
+}
+
+/// Future for the `into_sink` combinator, flattening a future-of-a-sink to
+/// get just the result of the final sink instead
+///
+/// This is created by the `Future::into_sink` method.
+pub struct FlattenSink<F: Future>
+    where F::Item: Sink
+{
+    state: State<F>,
+}
+
+pub fn new<F>(future: F) -> FlattenSink<F>
+    where F: Future,
+          F::Item: Sink
+{
+    FlattenSink {
+        state: State::Future(future),
+    }
+}
+
+impl<F: Future> Sink for FlattenSink<F>
+    where F::Item: Sink,
+          <F::Item as Sink>::SinkError: From<F::Error>,
+{
+    type SinkItem = <F::Item as Sink>::SinkItem;
+    type SinkError = <F::Item as Sink>::SinkError;
+    fn start_send(&mut self, item: Self::SinkItem)
+        -> StartSend<Self::SinkItem, Self::SinkError>
+    {
+        let (res, state) = match mem::replace(&mut self.state, State::Void) {
+            State::Future(mut fut) => {
+                match fut.poll() {
+                    Ok(Async::Ready(mut sink)) => {
+                        (sink.start_send(item)?, State::Sink(sink))
```

Comment:
Oh right yeah we shouldn't ourselves call the underlying sink again as part of normal usage, but I figure that we'd delegate to the underlying sink to "act improperly", if at all. In that sense this future would be robust in the face of use-after-error if the underlying sink was as well.

I do think though that if the future returns the error then this implementation should just be fused to forever return `NotReady`